feat(solana): add support for priority fees

gabrielKerekes opened the pull request:
Fixes https://github.com/trezor/trezor-suite/issues/11518.

Default priority fee should 5000 lamports - `50_000` compute units limit * `100_000` micro-lamports per unit. If the network is congested the fee (namely the price per unit) should be even higher based on [the getRecentPrioritizationFees RPC call](https://solana.com/docs/rpc/http/getrecentprioritizationfees).

I haven't yet been able to test this properly with a congested network (i.e. transaction passes on this branch but not on production). I'll see if I can do that tonight as it seems the network activity is highest in the evenings.

memoriesadrift commented on packages/blockchain-link/src/workers/solana/fee.ts:
```diff
@@ -0,0 +1,37 @@
+import BigNumber from 'bignumber.js';
+
+const DEFAULT_COMPUTE_UNIT_PRICE = 100_000; // micro-lamports, value taken from other wallets
+const DEFAULT_COMPUTE_UNIT_LIMIT = 50_000; // sending tokens with token account creation requires ~28K units
```
IIRC, this is fine, but while doing some reading I came across [this article](https://solana.com/developers/guides/advanced/how-to-use-priority-fees) where solana recommends "To best optimize our transaction, request exactly <number> CU with the Compute Budget Program when adding additional priority fees."
I'm not sure what the reason is behind that as I think only as many CU as needed are used anyway. 
Could you quickly test this?

gabrielKerekes replied:
> I think only as many CU as needed are used anyway

If you reserve 500K CUs, you pay for 500K CUs even if your transaction only uses 500.

Thanks for the comment though, because it pushed to revisit this and in the end I also implemented transaction simulation into the priority fee computations and so the limit will be exactly what the transaction needs + 20%. [This article](https://solana.com/developers/guides/advanced/how-to-request-optimal-compute) mentions a buffer of 10%, so we add 20% just to be sure.

However I'm not 100% convinced about this approach since e.g. SOL transfer now adds 18 lamports as fee i.e. the final fee is 5018 lamports instead of 5000 and I'm not entirely sure that's enough :D I'll discuss with Trezor guys, the previous approach is still implemented if we drop the last two commits.

memoriesadrift commented on packages/blockchain-link/src/workers/solana/fee.ts:
```diff
@@ -0,0 +1,37 @@
+import { Connection, Message, PublicKey } from '@solana/web3.js';
+import BigNumber from 'bignumber.js';
+
+const DEFAULT_COMPUTE_UNIT_PRICE = 100_000; // micro-lamports, value taken from other wallets
+const DEFAULT_COMPUTE_UNIT_LIMIT = 50_000; // sending tokens with token account creation requires ~28K units
+
+export const getBaseFee = async (api: Connection, message: string) => {
+    const result = await api.getFeeForMessage(Message.from(Buffer.from(message, 'hex')));
+    // The result can be null, for example if the transaction blockhash is invalid.
+    // In this case, we should fall back to the default fee.
+    if (result.value == null) {
+        throw new Error('Could not estimate fee for transaction.');
+    }
+
+    return result.value;
+};
+
+export const getPriorityFee = async (api: Connection, affectedAccounts: PublicKey[]) => {
+    const recentFees = await api.getRecentPrioritizationFees({
+        lockedWritableAccounts: affectedAccounts.map(a => new PublicKey(a)),
+    });
+
+    const networkPriorityFee = recentFees.map(a => a.prioritizationFee).sort((a, b) => b - a)[
+        Math.floor(recentFees.length / 4)
+    ]; // 25th percentile because many 0 priority fees are expected
+
+    const computeUnitPrice = new BigNumber(
+        Math.max(networkPriorityFee, DEFAULT_COMPUTE_UNIT_PRICE),
+    );
```
Don't we want to `Math.min` here to save the user fees if the network isn't congested? 
i.e. if all the recent priority fees are 0, as is often the case on devnet, I'd expect to not have to pay any priority fees.

gabrielKerekes replied:
No, we want to always set a priority fee just to be sure the transactions go through. Same is true for other wallets as well.

memoriesadrift commented on packages/blockchain-link/src/workers/solana/fee.ts:
```diff
@@ -0,0 +1,37 @@
+    const result = await api.getFeeForMessage(Message.from(Buffer.from(message, 'hex')));
+    // The result can be null, for example if the transaction blockhash is invalid.
+    // In this case, we should fall back to the default fee.
+    if (result.value == null) {
+        throw new Error('Could not estimate fee for transaction.');
+    }
```
Since this throws when the value is null, the priority fee will never get calculated in `blockchainEstimateFee`. Do you think that should happen or do we want to add a default priority fee on top in that case too?

gabrielKerekes replied:
There is a [default priority fee configured](https://github.com/vacuumlabs/trezor-suite/pull/34/commits/33c5bc78b9c899bf19b14a51f140461af0689507) which will be used if this throws. I'm not sure exactly how it works in the background but it does.

memoriesadrift reviewed (commented):
Some small changes, otherwise LGTM

gabrielKerekes committed:
fix(connect): add missing isCreatingAccount param validation in `blockchainEstimateFee`

gabrielKerekes committed:
feat(blockchain-link): add support for Solana priority fees

gabrielKerekes committed:
feat(connect): set default Solana fees

gabrielKerekes committed:
feat(suite): add support for Solana priority fees

gabrielKerekes committed:
fix(blockchain-link): fix Solana transaction confirmation

Combine `signatureSubscribe` and `getSignatureStatus` when confirming transactions. This resolves the issue with Quicknode not sending a signature update event. Transaction will confirm if either a signature update event is received or if the periodic `getSignatureStatus` confirms the transaction as `finalized`.

gabrielKerekes commented:
Merged via https://github.com/trezor/trezor-suite/pull/11703.
